File: src/main/java/com/github/pauldambra/moduluschecker/SortCodeSubstitution.java
package com.github.pauldambra.moduluschecker;

import com.github.pauldambra.moduluschecker.Account.BankAccount;
import com.google.common.base.CharMatcher;
import com.google.common.base.Charsets;
import com.google.common.base.Splitter;
import com.google.common.collect.ImmutableMap;
import com.google.common.io.Resources;

import java.io.IOException;
import java.net.URL;
import java.util.Map;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class SortCodeSubstitution {

    private static final Splitter newlineSplitter = Splitter.on(Pattern.compile("\r?\n"));
    private static final Splitter spaceSplitter = Splitter.on(CharMatcher.whitespace()).omitEmptyStrings();

    private final ImmutableMap<String, String> sortCodeSubstitutionTable;

    public SortCodeSubstitution(ImmutableMap<String, String> sortCodeSubstitutionTable) {
        this.sortCodeSubstitutionTable = sortCodeSubstitutionTable;
    }

    private SortCodeSubstitution(Map<String, String> sortCodeSubstitutionTable) {
        this.sortCodeSubstitutionTable = ImmutableMap.copyOf(sortCodeSubstitutionTable);
    }

    public BankAccount Apply(BankAccount bankAccount) {
        return sortCodeSubstitutionTable.containsKey(bankAccount.sortCode)
                ? BankAccount.Of(sortCodeSubstitutionTable.get(bankAccount.sortCode), bankAccount.accountNumber)
                : new BankAccount(bankAccount);
    }

    public static SortCodeSubstitution fromFile(String filePath) throws IOException {

        URL url = Resources.getResource(filePath);
        String text = Resources.toString(url, Charsets.UTF_8);

        Map<String, String> substitutions = newlineSplitter
                                                .splitToList(text)
                                                .stream()
                                                .map(spaceSplitter::splitToList)
                                                .filter(r -> r.size() != 0)
                                                .collect(Collectors.toMap(
                                                        ss -> ss.get(0),
                                                        ss -> ss.get(1)
                                                ));

        return new SortCodeSubstitution(substitutions);
    }
}

File: src/main/java/com/github/pauldambra/moduluschecker/SortCodeSubstitution.kt
package com.github.pauldambra.moduluschecker

import com.github.pauldambra.moduluschecker.account.BankAccount
import com.google.common.base.Charsets
import com.google.common.collect.ImmutableMap
import com.google.common.io.Resources
import java.io.IOException
import java.util.regex.Pattern

class SortCodeSubstitution {

    private val sortCodeSubstitutionTable: ImmutableMap<String, String>

    constructor(sortCodeSubstitutionTable: ImmutableMap<String, String>) {
        this.sortCodeSubstitutionTable = sortCodeSubstitutionTable
    }

    private constructor(sortCodeSubstitutionTable: Map<String, String>) {
        this.sortCodeSubstitutionTable = ImmutableMap.copyOf(sortCodeSubstitutionTable)
    }

    fun apply(bankAccount: BankAccount) =
      if (sortCodeSubstitutionTable.containsKey(bankAccount.sortCode))
        BankAccount.with(sortCodeSubstitutionTable[bankAccount.sortCode]!!, bankAccount.accountNumber)
      else
        BankAccount(bankAccount)

    companion object {

        @Throws(IOException::class)
        fun fromFile(filePath: String): SortCodeSubstitution {

            val url = Resources.getResource(filePath)
            val text = Resources.toString(url, Charsets.UTF_8)

            val substitutions = text.split(Pattern.compile("\r?\n"))
              .filter { it.isNotEmpty() }
              .map { it.split(' ') }
              .filter { it.isNotEmpty() }
              .associate { it[0] to it[1] }

            return SortCodeSubstitution(substitutions)
        }
    }
}
